### DISCUSSION TITLE
where to define USE_FULL_ASSERT ?

### DISCUSSION THREAD

**Discussion by PerennialNovice** [stm32duino/Arduino_Core_STM32#2374] [Q&A] [answered]
I tried to use this switch, but putting
`#define USE_FULL_ASSERT 1`
in hal_conf_extra.h does not seem to work

```
:\Sloeber\arduinoPlugin\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Src/stm32l4xx_hal_adc.c:409: undefined reference to `assert_failed'
c:/sloeber/arduinoplugin/packages/stmicroelectronics/tools/xpack-arm-none-eabi-gcc/12.2.1-1.2/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld.exe: C:\Sloeber\arduinoPlugin\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Src/stm32l4xx_hal_adc.c:410: undefined reference to `assert_failed'
c:/sloeber/arduinoplugin/packages/stmicroelectronics/tools/xpack-arm-none-eabi-gcc/12.2.1-1.2/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld.exe: C:\Sloeber\arduinoPlugin\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Src/stm32l4xx_hal_adc.c:411: undefined reference to `assert_failed'
c:/sloeber/arduinoplugin/packages/stmicroelectronics/tools/xpack-arm-none-eabi-gcc/12.2.1-1.2/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld.exe: C:\Sloeber\arduinoPlugin\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Src/stm32l4xx_hal_adc.c:413: undefined reference to `assert_failed'
c:/sloeber/arduinoplugin/packages/stmicroelectronics/tools/xpack-arm-none-eabi-gcc/12.2.1-1.2/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld.exe: C:\Sloeber\arduinoPlugin\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Src/stm32l4xx_hal_adc.c:415: undefined reference to `assert_failed'
c:/sloeber/arduinoplugin/packages/stmicroelectronics/tools/xpack-arm-none-eabi-gcc/12.2.1-1.2/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld.exe: .\libraries\SrcWrapper\src\HAL\stm32yyxx_hal_adc.c.o:C:\Sloeber\arduinoPlugin\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Src/stm32l4xx_hal_adc.c:416: more undefined references to `assert_failed' follow
collect2.exe: error: ld returned 1 exit status
make[1]: *** [makefile:165: LaserHead.elf] Error 1
make: *** [makefile:144: all] Error 2
```

What is the correct way of using this switch?

**Comment by fpistm** [comment]
Use build_opt.h.

**Comment by PerennialNovice** [comment]
thanks!

**Comment by PerennialNovice** [comment]
this does not seem to work...
on sloeber, nothing happens, USE_FULL_ASSERT remains undefined

on Arduino IDE 2.2.2
I get errors:

> C:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\xpack-arm-none-eabi-gcc\12.2.1-1.2/bin/arm-none-eabi-g++ -mcpu=cortex-m4 -mfpu=fpv4-sp-d16 -mfloat-abi=hard -DVECT_TAB_OFFSET=0x0 -DUSE_FULL_LL_DRIVER -mthumb @C:\Users\xXxXxXx\AppData\Local\Temp\arduino\sketches\47BAF4221CB4A2BA311E78AEDE5936AA/sketch/build.opt -c -Og -g -w -std=gnu++17 -ffunction-sections -fdata-sections -fno-threadsafe-statics --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -fno-use-cxa-atexit -IC:\Users\xXxXxXx\Documents\Arduino\Test -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/avr -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32 -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/LL -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/usb -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/OpenAMP -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/usb/hid -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/usb/cdc -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Inc -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Src -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/STM32L4xx -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/ST/STM32_USB_Device_Library/Core/Inc -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/ST/STM32_USB_Device_Library/Core/Src -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP/open-amp/lib/include -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP/libmetal/lib/include -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP/virtual_driver -w -x c++ -E -CC -DSTM32L4xx -DARDUINO=10607 -DARDUINO_NUCLEO_L4R5ZI -DARDUINO_ARCH_STM32 -DBOARD_NAME="NUCLEO_L4R5ZI" -DVARIANT_H="variant_NUCLEO_L4R5ZI.h" -DSTM32L4R5xx -DHAL_UART_MODULE_ENABLED -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\CMSIS\5.9.0/CMSIS/Core/Include/ -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/CMSIS/Device/ST/STM32L4xx/Include/ -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/CMSIS/Device/ST/STM32L4xx/Source/Templates/gcc/ -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\CMSIS\5.9.0/CMSIS/DSP/Include -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\CMSIS\5.9.0/CMSIS/DSP/PrivateInclude -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\variants\STM32L4xx\L4R5Z(G-I)T_L4R7ZIT_L4S5ZIT_L4S7ZIT -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\libraries\SrcWrapper\src C:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\libraries\SrcWrapper\src\stm32\uart.c -o nul
> Error while detecting libraries included by C:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\libraries\SrcWrapper\src\stm32\uart.c
> C:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\xpack-arm-none-eabi-gcc\12.2.1-1.2/bin/arm-none-eabi-g++ -mcpu=cortex-m4 -mfpu=fpv4-sp-d16 -mfloat-abi=hard -DVECT_TAB_OFFSET=0x0 -DUSE_FULL_LL_DRIVER -mthumb @C:\Users\xXxXxXx\AppData\Local\Temp\arduino\sketches\47BAF4221CB4A2BA311E78AEDE5936AA/sketch/build.opt -c -Og -g -w -std=gnu++17 -ffunction-sections -fdata-sections -fno-threadsafe-statics --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -fno-use-cxa-atexit -IC:\Users\xXxXxXx\Documents\Arduino\Test -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/avr -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32 -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/LL -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/usb -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/OpenAMP -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/usb/hid -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/usb/cdc -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Inc -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Src -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/STM32L4xx -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/ST/STM32_USB_Device_Library/Core/Inc -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/ST/STM32_USB_Device_Library/Core/Src -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP/open-amp/lib/include -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP/libmetal/lib/include -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP/virtual_driver -w -x c++ -E -CC -DSTM32L4xx -DARDUINO=10607 -DARDUINO_NUCLEO_L4R5ZI -DARDUINO_ARCH_STM32 -DBOARD_NAME="NUCLEO_L4R5ZI" -DVARIANT_H="variant_NUCLEO_L4R5ZI.h" -DSTM32L4R5xx -DHAL_UART_MODULE_ENABLED -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\CMSIS\5.9.0/CMSIS/Core/Include/ -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/CMSIS/Device/ST/STM32L4xx/Include/ -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/CMSIS/Device/ST/STM32L4xx/Source/Templates/gcc/ -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\CMSIS\5.9.0/CMSIS/DSP/Include -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\CMSIS\5.9.0/CMSIS/DSP/PrivateInclude -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\variants\STM32L4xx\L4R5Z(G-I)T_L4R7ZIT_L4S5ZIT_L4S7ZIT -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\libraries\SrcWrapper\src C:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\libraries\SrcWrapper\src\syscalls.c -o nul
> Error while detecting libraries included by C:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\libraries\SrcWrapper\src\syscalls.c
> Generating function prototypes...
> C:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\xpack-arm-none-eabi-gcc\12.2.1-1.2/bin/arm-none-eabi-g++ -mcpu=cortex-m4 -mfpu=fpv4-sp-d16 -mfloat-abi=hard -DVECT_TAB_OFFSET=0x0 -DUSE_FULL_LL_DRIVER -mthumb @C:\Users\xXxXxXx\AppData\Local\Temp\arduino\sketches\47BAF4221CB4A2BA311E78AEDE5936AA/sketch/build.opt -c -Og -g -w -std=gnu++17 -ffunction-sections -fdata-sections -fno-threadsafe-statics --param max-inline-insns-single=500 -fno-rtti -fno-exceptions -fno-use-cxa-atexit -IC:\Users\xXxXxXx\Documents\Arduino\Test -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/avr -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32 -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/LL -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/usb -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/OpenAMP -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/usb/hid -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino/stm32/usb/cdc -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Inc -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/STM32L4xx_HAL_Driver/Src -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/STM32L4xx -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/ST/STM32_USB_Device_Library/Core/Inc -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/ST/STM32_USB_Device_Library/Core/Src -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP/open-amp/lib/include -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP/libmetal/lib/include -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Middlewares/OpenAMP/virtual_driver -w -x c++ -E -CC -DSTM32L4xx -DARDUINO=10607 -DARDUINO_NUCLEO_L4R5ZI -DARDUINO_ARCH_STM32 -DBOARD_NAME="NUCLEO_L4R5ZI" -DVARIANT_H="variant_NUCLEO_L4R5ZI.h" -DSTM32L4R5xx -DHAL_UART_MODULE_ENABLED -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\CMSIS\5.9.0/CMSIS/Core/Include/ -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/CMSIS/Device/ST/STM32L4xx/Include/ -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\system/Drivers/CMSIS/Device/ST/STM32L4xx/Source/Templates/gcc/ -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\CMSIS\5.9.0/CMSIS/DSP/Include -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\tools\CMSIS\5.9.0/CMSIS/DSP/PrivateInclude -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\cores\arduino -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\variants\STM32L4xx\L4R5Z(G-I)T_L4R7ZIT_L4S5ZIT_L4S7ZIT -IC:\Users\xXxXxXx\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.7.1\libraries\SrcWrapper\src C:\Users\xXxXxXx\AppData\Local\Temp\arduino\sketches\47BAF4221CB4A2BA311E78AEDE5936AA\sketch\Test.ino.cpp -o C:\Users\xXxXxXx\AppData\Local\Temp\1719411759\sketch_merged.cpp
> arm-none-eabi-g++: warning: #define: linker input file unused because linking not done
> arm-none-eabi-g++: error: #define: linker input file not found: No such file or directory
> arm-none-eabi-g++: warning: USE_FULL_ASSERT: linker input file unused because linking not done
> arm-none-eabi-g++: error: USE_FULL_ASSERT: linker input file not found: No such file or directory
> arm-none-eabi-g++: warning: 1: linker input file unused because linking not done
> arm-none-eabi-g++: error: 1: linker input file not found: No such file or directory
> exit status 1
> 
> Compilation error: exit status 1

**Reply by fpistm** [reply]
First we do not support sloeber.
About Arduino, I've tested and do not have same issue.
First issue is that HAL try to include `stm32_assert.h` which does not exist as never implemented.
In `build_opt.h` you should define it like:
`-DUSE_FULL_ASSERT`

**Comment by PerennialNovice** [comment]
> > First we do not support sloeber.

I know, that is why I tried it in the official Arduino IDE.

> First issue is that HAL try to include stm32_assert.h which does not exist as never implemented.

OK, so there is no way to make use of all those assert-checks sprinkled along the source code
good to know before one tries :)

> -DUSE_FULL_ASSERT

ah, I wasn't aware of the correct syntax to use, thanks

**Reply by fpistm** [reply]
No worry :)
Could you file an issue about this support, please?

**Reply by fpistm** [reply]
Closed in favor of #2375

**Reply by fpistm** [reply, marked as answer]
Implemented by #2384.
Documentation: https://github.com/stm32duino/Arduino_Core_STM32/wiki/HAL-configuration#hal-assertion-management
